Homework 3 - Testing my application

fdemchenko opened the pull request:
In this homework I was trying to cover my application with test (unit test, integration test).
The most important thing I learned, as a person who wasn't writing tests too much before - high-quality test coverage requires good architecture and SOLID principles implemented. 

So, what i did

- write unit tests for validator package
- write unit test for email service and exchange rate service (together with refactoring it)
- write integration test for whole `/rate` endpoint

As always, any recommendations are welcome:)

Also I run simple load test with 10000 requests, using [hey](https://github.com/rakyll/hey)

![Screenshot 2024-06-16 195025](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/assets/89973537/7c615895-4c51-4797-b2d4-8b28c5e0e9f7)

![Screenshot 2024-06-16 195227](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/assets/89973537/c7982617-330b-4439-96cf-ebf0983b5a5f)

fdemchenko committed:
Write unit tests for validator

fdemchenko committed:
Create middleware that sets some security related headers

fdemchenko committed:
Test secureHeaders middleware

fdemchenko committed:
Test email service

fdemchenko committed:
Refactor rate service code

Add one more abstraction layer to rate service. Now rate service accept
rate client as a parameter, which will fecth exchange rate.
Also, change rate type from *Rate to float32, because *Rate is too
complex, and used only to unmarshal json object

fdemchenko committed:
Fix linter issues

fdemchenko committed:
Write unit test for caching rate service

fdemchenko committed:
Add /rate endpoint integration test

DimaL-cloud commented:
Written unit, integration and load tests. Great job. 🚀
I would suggest to try using test containers to write integration tests for services that use repositories. Also it is more easier to write integration tests than unit because you don't have to mock a lot of staff

fdemchenko committed:
Add email service integration test with test containers

MaxDMT16 commented on cmd/web/middleware.go:
```diff
@@ -25,3 +25,14 @@ func (app *application) RecoveryMiddleware(next http.Handler) http.Handler {
 	})
 }
+
+func secureHeaders(next http.Handler) http.Handler {
```
tiny: consider using consistent style for all middlewares.
`secureHeaders` is declared as function and other middlewares as methods of `app` struct

fdemchenko replied:
`app` struct is used as a dependencies provider in handlers level. Because `secureHeaders` does not need any dependencies from `app` it's declared as standalone function, but for consistency it's good idea to make it method of `app`.

fdemchenko replied:
Resolved with [227f2b9](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/227f2b91220cc70c3065a40d7bdb909270104116)

MaxDMT16 commented on cmd/web/middleware_test.go:
```diff
@@ -0,0 +1,38 @@
+	response := recorder.Result()
+	assert.Equal(t, response.StatusCode, http.StatusOK)
+
+	assert.Equal(t, "no-store", response.Header.Get("Cache-Control"))
+	assert.Equal(t, "frame-ancestors 'none'", response.Header.Get("Content-Security-Policy"))
+	assert.Equal(t, "nosniff", response.Header.Get("X-Content-Type-Options"))
+	assert.Equal(t, "DENY", response.Header.Get("X-Frame-Options"))
```
what about other headers?)

advice: you could make the implementation more flexible by having a map of headers and their values. Then in test you could iterate over the map and check all headers. In case new header will be added, no changes will be needed in test cause it will reference the same map with headers

fdemchenko replied:
Totally agree!

fdemchenko replied:
Resolved with [227f2b9](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/227f2b91220cc70c3065a40d7bdb909270104116)

MaxDMT16 commented on cmd/web/server_test.go:
It would be great to add integration test on /subscribe endpoint

MaxDMT16 commented on cmd/web/server_test.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	_, err = strconv.ParseFloat(string(body), 32)
+	assert.NoError(t, err)
```
I found useful to add a check the value is greater than `0`

fdemchenko replied:
Resolved with [b8ee5a7](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/b8ee5a7212e74c281d914af6ec9f5f8fa1eb91e9)

MaxDMT16 commented on internal/database/helpers.go:
I'm glad to see that you continue improving your project, besides of HA tasks)
Well done!

MaxDMT16 replied:
However, name of the file is inappropriate I think.
Try to avoid names like `helper`/`util`. They are to general.

By default in go if package has a single file inside, it's usually named as the name of the package. But if there is more appropriate name of the file, it's better to use it.

In this case, `migration.go` seems to be more descriptive

fdemchenko replied:
Resolved with [ab00bd4](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/ab00bd4e73ca832d380d13972f485416c91c6412)

MaxDMT16 commented on internal/integration/email_test.go:
```diff
@@ -0,0 +1,89 @@
+	}
+}
+
+func (em *EmailServiceSuite) TestCreateEmail_Success() {
+	err := em.emailService.Create("someemail@gmail.com")
+	assert.NoError(em.T(), err)
+}
+
+func (em *EmailServiceSuite) TestCreateEmail_Duplicate() {
+	t := em.T()
+	err := em.emailService.Create("somemail@gmail.com")
+	assert.NoError(t, err)
+
+	err = em.emailService.Create("somemail@gmail.com")
+	assert.ErrorIs(t, err, repositories.ErrDuplicateEmail)
+}
```
These tests can be merged into one with two test cases.
Consider using [table test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) in such cases

MaxDMT16 commented on internal/integration/setup.go:
```diff
@@ -0,0 +1,49 @@
+	}
+	_, _, err = postgresContainer.Exec(ctx, []string{"psql", "-U", dbUser,
+		"-d", dbName,
+		"-c", "CREATE TABLE emails (id SERIAL PRIMARY KEY, email TEXT UNIQUE NOT NULL)"})
```
If table structure will change you have to add a migration script and fix script in test.
It's better to unify this process and make test container to apply the same migrations that app does when starting

fdemchenko replied:
Absolutely agree. This workaround is going to create more problems in the future. It's temporary solution, will change it with next commit.

fdemchenko replied:
Resolved with [dd0b305](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/dd0b305bea1a7ac8419f040898841fdf188ce356). Now, migrations applying to test container.

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -23,18 +23,26 @@ type Rate struct {
 
-func NewRateService(updateInterval time.Duration) *CachingRateService {
+// Creates new rate service instance.
+// Pass nil to client to use default http client.
```
tiny: For such cases [options pattern](https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c) is used in Go

fdemchenko replied:
Thanks a lot! Will use it in the next home task.

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -68,29 +76,37 @@ func (rs *CachingRateService) StartBackgroundTask() {
 
-func (rs *CachingRateService) fetchExchangeRate() (*Rate, error) {
-	resp, err := http.Get(RatesAPIBaseURL + "/usd.json")
+type HTTPExchangeRateClient struct {
```
Good improvement 👍

MaxDMT16 replied:
Is it expected this client will be instantiated without constructor (`NewHTTPExchangeRateClient`)?
If it's not, I don't see a reason to make this struct available outside of the package.

P.S. the same comment to other structs in the repo)

fdemchenko replied:
This topic is discussed here https://github.com/golang/lint/issues/210. And I agree, for better encapsulation I should make structs unexported, but revive linter gives me warning like this `unexported-return: exported func NewRateService returns unexported type *services.cachingRateService, which can be annoying to use`. So I will just ignore these warnings for now.

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -68,29 +76,37 @@ func (rs *CachingRateService) StartBackgroundTask() {
+	return &HTTPExchangeRateClient{client: client}
+}
+
+func (ec *HTTPExchangeRateClient) GetExchangeRate() (float32, error) {
```
Do you mind making client flexible to be able return exchange rate of different currency pairs?

fdemchenko replied:
Yes, I'm going to extend rate clients capabilities in the next homework, as well as creating several additional exchange rate providers.

MaxDMT16 commented on internal/services/rate_test.go:
```diff
@@ -0,0 +1,42 @@
+	return args.Get(0).(float32), args.Error(1)
+}
+
+func TestRateService(t *testing.T) {
```
Here you're testing several things at the same time:
- if fetched rate is correct
- if rate is cached
- if rate is re-fetched after update interval

These are different act parts from AAA approach.
I'd suggest implementing it as different tests.
Then tests will become more clear and easier to maintain

fdemchenko replied:
Resolved with [95ae405](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/95ae405cbefc92d7ec1edb2d1f12c7e469b05e9b)

MaxDMT16 commented on internal/validator/validator_test.go:
```diff
@@ -0,0 +1,51 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidator_ValidEmails(t *testing.T) {
```
Nice, you have table tests!

MaxDMT16 replied:
tiny: I think these tests can be merged into one. Than test case struct will look like
```
{
  name  string
  email string
  isValid bool
  expectedErrorMessage string
}
```

fdemchenko replied:
Resolved with [c4f7ffb](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/c4f7ffb0095e9a0b4768fbf18e85014cd109b08a)

MaxDMT16 reviewed (approved):
LGTM in general
Most comments are small suggestions or advices

MaxDMT16 commented:
Also it would be great to update readme file with info about tests (what test types are present ing the project and how to run them)

fdemchenko committed:
Bring some consistency to middlewares. Create secure headers map

fdemchenko committed:
Separate tests of rate service

fdemchenko committed:
Remove not descriptive file name, change from helpers.go to migrations.go

fdemchenko committed:
Combine validator tests into one

fdemchenko committed:
Add check for rate to be greater than zero

fdemchenko committed:
Fix error with rate type

fdemchenko committed:
Reorginize database migrations

Embed migrations directory into application.
Add ability to close connection after migrations happened (primarily for
tests)
Remove copying migrations directive from Dockerfile

fdemchenko committed:
Apply migrations for test database container

fdemchenko committed:
Make struct instaniated from constructors private

fdemchenko committed:
Ignore revive rule for creating unexported structs with public constructors
